### DISCUSSION TITLE
Why isn't the latest version of Kotlin being found in this case?

### DISCUSSION THREAD

**Discussion by xenomachina** [ben-manes/gradle-versions-plugin#823] [Q&A] [answered]
The snippets below are pared down from a gradle plugin we have that in turn
relies on the kotlin-gradle-plugin. For some reason dependencyUpdates isn't
detecting when the kotlin dependency is out of date. I found a workaround, but
I don't really understand why it works or why it's necessary.

`gradle.properties` contains:

```
kotlin_version = 1.6.0
```

and `build.gradle` contains:
``` groovy
buildscript {
    dependencies {
        classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
    }
}

plugins {
    id 'com.github.ben-manes.versions' version '0.50.0'
    id 'java-gradle-plugin'
}

repositories {
    mavenCentral()
}

dependencies {
    implementation "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
}
```

If I run `./gradlew dependencyUpdates`, it says:

```
The following dependencies are using the latest milestone version:
 - com.github.ben-manes.versions:com.github.ben-manes.versions.gradle.plugin:0.50.0
 - org.jetbrains.kotlin:kotlin-gradle-plugin:1.6.0
```

However, if I add the following to the `plugins` block in `build.gradle`:

``` groovy
id 'org.jetbrains.kotlin.jvm' version "$kotlin_version"
```

and then run `./gradlew dependencyUpdates`, it says:

```
The following dependencies have later milestone versions:
 - org.jetbrains.kotlin.jvm:org.jetbrains.kotlin.jvm.gradle.plugin [1.6.0 -> 2.0.0-Beta2]
     https://kotlinlang.org/
```

So addiing the plugin dependency "fixes" the issue, but why is this necessary
for `dependencyUpdates` to see that there are newer versions of Kotlin?
The code builds and functions correctly without it.

(I'm using Gradle 8.1.1, in case it matters.)

**Comment by ben-manes** [comment]
I'm not sure. If I drop down a few plugin versions, e.g. compare v46 and v47, then it shows up. Can you confirm?

If so, then the only change that looks relevant came from the Gradle team (https://github.com/ben-manes/gradle-versions-plugin/pull/768)

**Reply by xenomachina** [reply]
Yes, I'm seeing that behavior as well. With 0.0.47:

```
The following dependencies are using the latest milestone version:
 - org.jetbrains.kotlin:kotlin-gradle-plugin:1.6.0
 - org.jetbrains.kotlin:kotlin-stdlib:1.6.0
```

but with 0.46.0:

```
The following dependencies have later milestone versions:
...
- org.jetbrains.kotlin:kotlin-gradle-plugin [1.6.0 -> 2.0.0-Beta2]
     https://kotlinlang.org/
 - org.jetbrains.kotlin:kotlin-stdlib [1.6.0 -> 2.0.0-Beta2]
     https://kotlinlang.org/
```

**Reply by ben-manes** [reply]
@tresat @jvandort do you have theories on this regression and recommendations for next steps? It looks like your change caused this, but I doubt reverting is the right fix here. It seems kind of nuanced.

**Reply by xenomachina** [reply]
I just noticed an error in my original post: this line wasn't supposed to be in the `plugins` block in the first snippet for `build.gradle`:

``` groovy
id 'org.jetbrains.kotlin.jvm' version "$kotlin_version"
```

(I think the later context may have clarified this, but I edited the post to correct this.)

**Reply by xenomachina** [reply]
Unfortunately, it looks like adding that dependency doesn't actually work as well as I'd hoped. :disappointed: 

When I add the plugin dependency, it says:

```
The following dependencies have later milestone versions:
 - org.jetbrains.kotlin:kotlin-gradle-plugin [1.6.0 -> 2.0.0-Beta2]
     https://kotlinlang.org/
 - org.jetbrains.kotlin.jvm:org.jetbrains.kotlin.jvm.gradle.plugin [1.6.0 -> 2.0.0-Beta2]
     https://kotlinlang.org/
```

which I *thought* was the correct behavior, but I'd missed the fact that it *also* says:

```
The following dependencies are using the latest milestone version:
...
 - org.jetbrains.kotlin:kotlin-compiler-embeddable:1.6.0
 - org.jetbrains.kotlin:kotlin-klib-commonizer-embeddable:1.6.0
```

**Reply by ben-manes** [reply]
A blind guess is that since we resolve a copy of the configuration with a dynamic version any resolution strategies, constraints, etc. are also applied. It may be that Kotlin is constraining the dynamic version (`+`), e.g. a BOM to force the version being used. That is sometimes in the Gradle info log but isn't always clear.

**Reply by xenomachina** [reply]
Hmmm... looks like my "workaround" not only didnt' work, it actually kind of backfires.

With `build.gradle`:

``` groovy
plugins {
    id 'com.github.ben-manes.versions' version '0.46.0'
    //id 'org.jetbrains.kotlin.jvm' version "$kotlin_version"
    id 'java-gradle-plugin'
}

repositories {
    mavenCentral()
}

dependencies {
    implementation "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
    implementation "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
}
```

and `gradle.properties`:

``` properties
kotlin_version = 1.8.22
```

`./gradlew dependencyUpdates` works correctly.

However, if I uncomment the `org.jetbrains.kotlin.jvm` dependency it incorrectly says:

```
The following dependencies are using the latest milestone version:
 - org.jetbrains.kotlin:kotlin-compiler-embeddable:1.8.22
 - org.jetbrains.kotlin:kotlin-klib-commonizer-embeddable:1.8.22
```

despite also seeing that some Kotlin artifacts have newer versions.

**Reply by xenomachina** [reply]
Ah, that might be it. I remember running into [a similar problem with the jooq gradle plugin](https://github.com/etiennestuder/gradle-jooq-plugin/issues/162). We ended up conditionally applying that plugin to work around this.

**Reply by ben-manes** [reply]
I don’t know why plug-ins try to force versions using resolution rules, but it’s a common problem. I believe the intended approach is to use a [default dependency](https://docs.gradle.org/current/userguide/implementing_gradle_plugins.html#providing_default_dependencies_for_plugins) so that it is a softer version that doesn’t interact badly with the rest of the build.

**Reply by xenomachina** [reply]
Yeah, I wish plugins wouldn't mess with the resolver.

In any case, this works correctly with 0.46.0, but not with 0.50.0. Is that a bug in 0.50.0? Should an issue be filed about this?

`build.gradle`:

``` groovy
plugins {
    id 'com.github.ben-manes.versions' version '0.46.0'
    id 'java-gradle-plugin'
}

repositories {
    mavenCentral()
}

dependencies {
    implementation "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
    implementation "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
}
```

`gradle.properties`:

``` properties
kotlin_version = 1.8.22
```

**Reply by ben-manes** [reply]
You’re welcome to open an issue with your reproduced, but please tag the Gradle folks. The pr mentioned above is the likely culprit and I’m not sure why. I don’t know when I’ll get a chance to investigate it though. I personally use both this plugin and dependabot for Caffeine to simplify upgrading, etc

**Reply by xenomachina** [reply]
What is "dependabot for Caffeine"?

Our code is on GitLab, not Github, and as far as I can tell, dependabot doesn't work on GitLab. We have our own in-house tool that relies on gradle-versions-plugin and gradle-use-latest-versions-plugin for our Gradle dependencies.

**Reply by ben-manes** [reply]
Oh sorry, I just meant my other [open-source project](https://github.com/ben-manes/caffeine) which is on github.

**Reply by xenomachina** [reply]
Ah, gotcha! I totally mis-parsed that as something like "Windows Subsystem for Linux". :laughing:

**Reply by tresat** [reply, marked as answer]
This issue certainly seems related to that linked change, and how the plugin isn't building a query dependency for the Kotlin plugin.  I'll take a closer look next week.
